clang/lib/Sema/SemaType.cpp

```diff
@@ -9267,14 +9267,14 @@ bool Sema::RequireLiteralType(SourceLocation Loc, QualType T,
   if (!RT)
     return true;
 
-  const CXXRecordDecl *RD = cast<CXXRecordDecl>(RT->getDecl());
-
   // A partially-defined class type can't be a literal type, because a literal
   // class type must have a trivial destructor (which can't be checked until
   // the class definition is complete).
   if (RequireCompleteType(Loc, ElemType, diag::note_non_literal_incomplete, T))
     return true;
 
+  const CXXRecordDecl *RD = cast<CXXRecordDecl>(RT->getDecl());
```

Comment:
I am surprised we have not run into this before this code was added in 2012.

Reply:
> I am also curious why this does not show up in C++, we obtain a similar diagnostic but no crash.

Please correct me if I'm wrong, the `CXXRecordDecl` is specific to C++ declarations, so the cast works as expected. In a C env, `RecordDecl` is used instead. We could wrap this in a condition to check specifically for C++, but `*RD` isn't required for `RequireCompleteType`. `RequireCompleteType` handles cases related to incomplete C types, so I'm not sure if we need to perform additional checks here.

```cpp
if (RequireCompleteType(Loc, ElemType, diag::note_non_literal_incomplete, T))
    return true;
```

Reply:
@MitalAshok Thanks for the feedback. `RequireLiteralType` is called directly in the condition. Should we perform additional checks directly in the condition, or should we include them within `RequireLiteralType` to mark the declaration as invalid?

Reply:
You don't require a literal type for C23 `constexpr`, so part of the condition should be `if (getLangOpts().CPlusPlus && NewVD->isConstexpr() && ... && RequireLiteralType`, just like how `CheckC23ConstexprVarType` is gated behind `getLangOpts().C23 &&`. And add an assertion for `getLangOpts().CPlusPlus` in RequireLiteralType for good measure.

The declaration will be marked invalid later by the same code that marks `/*!constexpr*/ struct Incomplete s = {};` invalid.

Reply:
@MitalAshok Oke, I've added changes. BTW, should I move const `CXXRecordDecl *RD` back to its original position after `RequireCompleteType`? Thanks

Reply:
> should I move const CXXRecordDecl *RD back to its original position after RequireCompleteType?

I think yes. It also makes sense to add `assert(LangOpts.CPlusPlus)` in the beginning of `RequireLiteralType` as @MitalAshok already mentioned.

Reply:
@Fznamznon Thanks. I've reverted. Yep, this isn't necessary to fix the issue., it's more about code cleanup.

Reply:
@shafik @MitalAshok Could you review the latest changes? If there are no further comments, could you please proceed with the merge as I don't have access to do so? Thank you!

Reply:
@shafik @MitalAshok Could you review the latest changes? If there are no further comments, could you please proceed with the merge as I don't have access to do so? Thank you!